classes/privacy/provider.php

```diff
@@ -0,0 +1,266 @@
+ *
+ * Implement support for describing and managing private data.
+ *
+ * @package    block_kuracloud
+ * @category   privacy
+ * @copyright  2020 KuraCloud
+ * @author     John Gee
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace block_kuracloud\privacy;
+
+defined('MOODLE_INTERNAL') || die();
+
+use \core_privacy\local\metadata\collection;
+use \core_privacy\local\request\contextlist;
+use \core_privacy\local\request\writer;
+use \core_privacy\local\request\userlist;
+use \core_privacy\local\request\approved_contextlist;
+use \core_privacy\local\request\approved_userlist;
+
+
+/**
+ * Delete user data for target user which is no longer associated with a valid course.
+ *
+ * @param  int $userid The target used.
+ */
+function kuracloud_delete_stale_data_for_user(int $userid) {
+    global $DB;
+
+    // Get list of stale records, which are no longer associated with a course.
+    $sql = "SELECT kc_users.id
+              FROM {block_kuracloud_users} kc_users
+         LEFT JOIN {block_kuracloud_courses} kc_courses on kc_users.remote_instanceid = kc_courses.remote_instanceid AND kc_users.remote_courseid = kc_courses.remote_courseid
+         LEFT JOIN {course} c on c.id = kc_courses.courseid
+             WHERE c.id IS NULL AND kc_users.userid = :userid
+    ";
+    $params = [
+        'userid' => $userid,
+    ];
+    $ids = $DB->get_fieldset_sql($sql, $params);
+
+    if (!empty($ids)) {
+        list($idinsql, $idinparams) = $DB->get_in_or_equal($ids, SQL_PARAMS_NAMED);
+        // Putting the userid in again for extra safety.
+        $params = array_merge(['userid' => $userid], $idinparams);
+        $sql = "userid = :userid AND id {$idinsql}";
+        $DB->delete_records_select('block_kuracloud_users', $sql, $params);
+    }
+}
+
+
+class provider implements
+    \core_privacy\local\metadata\provider,
+    \core_privacy\local\request\plugin\provider,
+    \core_privacy\local\request\core_userlist_provider {
+
+    public static function get_metadata(collection $collection) : collection {
+        // Describe the data we are storing locally.
+        $collection->add_database_table(
+            'block_kuracloud_users',
+            [
+                'userid' => 'privacy:metadata:block_kuracloud_users:userid',
+                'remote_studentid' => 'privacy:metadata:block_kuracloud_users:remote_studentid',
+
+            ],
+            'privacy:metadata:block_kuracloud_users'
+        );
+
+        // Describe the data exported to kuraCloud.
+        $collection->add_external_location_link('kuracloud_sync', [
+            'firstname' => 'privacy:metadata:kuracloud_sync:firstname',
+            'lastname' => 'privacy:metadata:kuracloud_sync:lastname',
+            'idnumber' => 'privacy:metadata:kuracloud_sync:idnumber',
+            'email' => 'privacy:metadata:kuracloud_sync:email',
+        ], 'privacy:metadata:kuracloud_sync');
+
+        return $collection;
+    }
```

Comment:
Why doesn't this one get a function comment?

Reply:
Huh. Will do.

Reply:
Added